third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -98,33 +110,76 @@ impl Interpreter {
         extensions: &mut NativeContextExtensions,
         loader: &Loader,
     ) -> VMResult<Vec<Value>> {
-        Interpreter {
-            operand_stack: Stack::new(),
-            call_stack: CallStack::new(),
-            paranoid_type_checks: loader.vm_config().paranoid_type_checks,
-            access_control: AccessControlState::default(),
-            active_modules: HashSet::new(),
-        }
-        .execute_main(
-            loader,
+        InterpreterImpl::entrypoint(
+            function,
+            args,
             data_store,
             module_store,
             module_storage,
             gas_meter,
             traversal_context,
             extensions,
-            function,
-            args,
+            loader,
         )
     }
+}
+
+impl InterpreterImpl {
+    /// Entrypoint into the interpreter. All external calls need to be routed through this
+    /// function.
+    pub(crate) fn entrypoint(
+        function: LoadedFunction,
+        args: Vec<Value>,
+        data_store: &mut TransactionDataCache,
+        module_store: &LegacyModuleStorageAdapter,
+        module_storage: &impl ModuleStorage,
+        gas_meter: &mut impl GasMeter,
+        traversal_context: &mut TraversalContext,
+        extensions: &mut NativeContextExtensions,
+        loader: &Loader,
+    ) -> VMResult<Vec<Value>> {
+        let interpreter = InterpreterImpl {
+            operand_stack: Stack::new(),
+            call_stack: CallStack::new(),
+            paranoid_type_checks: loader.vm_config().paranoid_type_checks,
+            access_control: AccessControlState::default(),
+            active_modules: HashSet::new(),
+        };
+
+        if loader.vm_config().paranoid_type_checks {
```

Comment:
This is beautiful! One thing to note: we also have some paranoid checks inside execution loop (e.g., for functions). These can also be moved to this generic, but maybe you have a separate PR for that?

Reply:
Exactly - following PRs